### PR TITLE
Add unpublish message feature

### PR DESCRIPTION
**Pull request by 1pretz1**
Now when a message is received by the email-alert-api that a taxon has been unpublished, the users who are subscribed to that taxon will receive an email notifying them that their subscription has ended. 

Currently only taxon's are being processed if they've been unpublished but the plan is to also add this feature for legacy taxons like Policy Areas and Specialist Sectors etc.

(Emails wont actually be sent as is, only logged)

Trello;
https://trello.com/c/fPAplaaV/67-send-notifications-when-unpublishing-taxons

### CODE REVIEW

File: config/routes.rb [hunk 1]
```diff
@@ -1,10 +1,14 @@
-Rails.application.routes.draw do # rubocop:disable Metrics/BlockLength
+# rubocop:disable Metrics/BlockLength
```
Comment (cbaines) [inline config/routes.rb]: I think it would be best to disable this here https://github.com/alphagov/email-alert-api/blob/master/.rubocop.yml as that avoids the distraction of the comments in the file.

Whitehall might serve as a good example https://github.com/alphagov/whitehall/blob/master/.rubocop.yml

File: spec/builders/unpublish_email_builder_spec.rb [hunk 2]
```diff
@@ -0,0 +1,65 @@
+# rubocop:disable Layout/EmptyLinesAroundArguments
```
Comment (cbaines) [inline spec/builders/unpublish_email_builder_spec.rb]: Same thing for this https://github.com/alphagov/email-alert-api/pull/655/review_comment/create#r209238087

File: app/controllers/unpublish_messages_controller.rb [hunk 3]
```diff
@@ -0,0 +1,20 @@
+class UnpublishMessagesController < ApplicationController
+  def create
+    UnpublishHandlerService.call(
+      unpublishing_params[:content_id]
+    )
+
+    render json: { message: "Unpublish message queued for sending" }, status: 202
+  end
+
+  def index
```
[the remark this answers is not in the document]
Comment (1pretz1) [reply app/controllers/unpublish_messages_controller.rb]: Just following the pattern of the `NotificationsController`, i can remove it however and just have the create action.

File: config/routes.rb (hunk 1 above)
Comment (1pretz1) [reply config/routes.rb]: Sure, thanks for the example!

File: spec/builders/unpublish_email_builder_spec.rb (hunk 2 above)
Comment (1pretz1) [reply spec/builders/unpublish_email_builder_spec.rb]: 👍

File: app/controllers/unpublish_messages_controller.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (koetsier) [reply app/controllers/unpublish_messages_controller.rb]: think so. had a look at the notification controller and the relevant commit for the index  says:

>Make deprecated notification endpoint a no-op  
>These used to talk to GovDelivery but now they won't return anything useful.

So I don't think it is necessary in this  case

[the remark this answers is not in the document]
Comment (1pretz1) [reply app/controllers/unpublish_messages_controller.rb]: Okay dokay, i've removed it and the corresponding route!

File: spec/builders/digest_email_builder_spec.rb [hunk 4]
```diff
@@ -132,5 +130,3 @@ def simulated_deduplification(content_changes)
     end
   end
 end
-
-# rubocop:enable Layout/EmptyLinesAroundArguments
```
Comment (cbaines) [inline spec/builders/digest_email_builder_spec.rb]: :+1:

File: spec/services/unpublish_handler_service_spec.rb [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+        .to change { Email.count }.by(2)
+      end
+
+     # it "increases the size of the queue in the DeliverRequestWorker" do
+     #   Sidekiq::Testing.fake! do
+     #     DeliveryRequestWorker.jobs.clear
+     #     described_class.call(@content_id)
+     #   end
+     #   expect(DeliveryRequestWorker.jobs.size).to eq(2)
+     # end
```
Comment (cbaines) [inline spec/services/unpublish_handler_service_spec.rb]: Could we mark this as pending somehow [1], rather than commenting it out?

1: https://relishapp.com/rspec/rspec-core/v/2-0/docs/pending/pending-examples

File: app/services/unpublish_handler_service.rb [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+        }
+      end
+    end
+  end
+
+  # For this query to return the content id has to be wrapped in a double quote blame psql 9.3
+  def subscriber_lists
+    @subscriber_lists ||= SubscriberList
+    .where(":id IN (SELECT json_array_elements((json_each(links)).value)::text)", id: "\"#{content_id}\"")
+    .includes(:subscribers).to_a
```
Comment (cbaines) [inline app/services/unpublish_handler_service.rb]: This layout might be a little more readable.
```ruby
    @subscriber_lists ||= SubscriberList.where(
      ":id IN (SELECT json_array_elements((json_each(links)).value)::text)", id: "\"#{content_id}\""
    ).includes(
      :subscribers
    ).to_a
```

Or maybe just indent the lines after SubscriberList by two more spaces?

File: app/services/unpublish_handler_service.rb [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+        id: #{email.id}
+        subject: #{email.subject}
+        body: #{email.body}
+        subscriber_id: #{email.subscriber_id}
+        ----
+      INFO
+    end
+  end
+
+  def log_none_taxon_lists(subscriber_lists)
```
Comment (cbaines) [inline app/services/unpublish_handler_service.rb]: I'm not sure what "none" means here, were you thinking of "non"?

File: app/services/unpublish_handler_service.rb [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+    subscriber_lists.each do |list|
+      logger.info(<<-INFO.strip_heredoc)
+        ++++
+        Email not created but matching a SubscriberList (Policy area etc):
```
Comment (cbaines) [inline app/services/unpublish_handler_service.rb]: Topics (taxons) use subscriber lists too, maybe say "Not sending notification about non-Topic SubscriberList".

File: spec/services/unpublish_handler_service_spec.rb [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+      end
+
+      it "doesn't create email if subscriber is deactivated" do
+        expect { described_class.call(@content_id) }.to_not(change { Email.count })
+      end
+    end
+
+    context "email is not created if there is no SubscriberList" do
+      it "doesn't build an email if there is no associated SubscriberList" do
+        content_id_two = "content-id"
```
Comment (cbaines) [inline spec/services/unpublish_handler_service_spec.rb]: Maybe `"non-existent-content-id"` to make it clear (assuming that is what's going on).

File: app/services/unpublish_handler_service.rb [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+  def initialize(content_id)
+    @content_id = content_id
+  end
+
+  def self.call(*args)
+    new(*args).call
+  end
+
+  def call
+    if email_params.empty?
```
Comment (koetsier) [inline app/services/unpublish_handler_service.rb]: email_params is also called in line 16 and could be memoized

File: app/services/unpublish_handler_service.rb [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+
+  def queue_delivery_request_workers(email_ids)
+    email_ids.each do |email_id|
+      DeliveryRequestWorker.perform_async_in_queue(
+        email_id, queue: :delivery_immediate
+      )
+    end
+  end
+
+  def email_params
```
Comment (koetsier) [inline app/services/unpublish_handler_service.rb]: I think sometimes it is clearer to explicitly pass in subscriber_lists so that you can see at the calling method that this depends on subscriber_lists. Same with some of the other methods

File: app/services/unpublish_handler_service.rb (hunk 10 above)
Comment (koetsier) [reply app/services/unpublish_handler_service.rb]: You could also make UnpublishEmailBuilder deal with an empty array

File: app/builders/unpublish_email_builder.rb [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+    BODY
+  end
+
+  def presented_manage_subscriptions_links(address)
+    ManageSubscriptionsLinkPresenter.call(address: address)
+  end
+
+  def courtesy_email
+    addresses = [
+      "govuk-email-courtesy-copies@digital.cabinet-office.gov.uk",
```
Comment (koetsier) [inline app/builders/unpublish_email_builder.rb]: can we make this a constant somewhere? I believe it is used elsewhere too

File: spec/services/unpublish_handler_service_spec.rb (hunk 9 above)
Comment (1pretz1) [reply spec/services/unpublish_handler_service_spec.rb]: Yup, that makes it clearer!

File: spec/services/unpublish_handler_service_spec.rb (hunk 5 above)
Comment (1pretz1) [reply spec/services/unpublish_handler_service_spec.rb]: Yep sure, thats interesting, didn't know you could do that.

File: app/services/unpublish_handler_service.rb (hunk 6 above)
Comment (1pretz1) [reply app/services/unpublish_handler_service.rb]: I've indented and moved the `.to_a` to a separate line which i think looks the best!

File: app/services/unpublish_handler_service.rb (hunk 7 above)
Comment (1pretz1) [reply app/services/unpublish_handler_service.rb]: Yep, got the two mixed up! -.-

File: app/services/unpublish_handler_service.rb (hunk 8 above)
Comment (1pretz1) [reply app/services/unpublish_handler_service.rb]: 👍

File: app/services/unpublish_handler_service.rb (hunk 10 above)
Comment (1pretz1) [reply app/services/unpublish_handler_service.rb]: Yep i could memoize email_params, however because I memoized `subscriber_lists` I assumed it would be okay to leave.

However i'll explore making `UnpublishEmailBuilder` handle an empty array instead, which will make this problem obsolete and probably make the code look nicer!

File: app/services/unpublish_handler_service.rb (hunk 11 above)
Comment (1pretz1) [reply app/services/unpublish_handler_service.rb]: I've done this for all the other methods, however i think for `email_params` it would look messy as i call `UnpublishEmailBuilder.call(email_params)` ? 

But noted that i should pass arguments more often.

Comment (koetsier) [reply app/services/unpublish_handler_service.rb]: no I meant def ```email_params(subscriber_lists)``` so you know what stuff this method depends on

Comment (1pretz1) [reply app/services/unpublish_handler_service.rb]: Yep i know, i meant that because i call the method from `UnpublishEmailBuilder.call(email_params)`, it looks a little messy to me to have `UnpublishEmailBuilder.call(email_params(subscriber_lists))` but if you think it would make things clearer then i'm not opposed to changing it :)

Comment (1pretz1) [reply app/services/unpublish_handler_service.rb]: I've added the arguments.

File: app/services/unsubscribe_service.rb [hunk 13]
```diff
@@ -1,37 +1,52 @@
...
-      unsubscribe!(subscriber, subscriber.active_subscriptions, :marked_as_spam, delivery_attempt.email)
+      ActiveRecord::Base.transaction do
+        unsubscribe_subscriptions!(subscriber.active_subscriptions, :marked_as_spam, delivery_attempt.email)
+        unsubscribe_subscriber!(subscriber)
+      end
     end
 
-  private
-
-    def unsubscribe!(subscriber, subscriptions, reason, email = nil)
```
Comment (cbaines) [inline app/services/unsubscribe_service.rb]: I'm a bit unclear what part of this commit is refactoring, and what purpose the refactoring is for. It would be useful to have some more information in the commit message, or maybe either separate the refactoring out in to a different commit.

**Comment by cbaines** [comment]
I've pulled out a rubocop change in `config/routes.rb` in to the `.rubocop.yml`, just to make the `config/routes.rb` file a bit cleaner.

File: app/services/unsubscribe_service.rb [hunk 14]
```diff
@@ -1,37 +1,52 @@
...
+    def subscriber_list!(list, reason)
       ActiveRecord::Base.transaction do
-        subscriptions.each do |subscription|
-          subscription.end(reason: reason)
+        subscriptions = list.subscriptions
+        unsubscribe_subscriptions!(subscriptions, reason)
+        Subscriber.where(subscriptions: subscriptions).each do |subscriber|
```
Comment (cbaines) [inline app/services/unsubscribe_service.rb]: Given this is happening in the request (as far as I can see), I don't know how this will cope with lists with many subscriptions. The largest I know of for the legacy things is ~16000, and I'd worry that doing 16000 queries inside the request won't work.

Not something we necessarily need to solve now, but if I'm right, it might be worth improving soon so that this will work with large lists.

Comment (koetsier) [reply app/services/unsubscribe_service.rb]: You're right - made this call asynchronous
